app/components/AppProvider.tsx

```diff
@@ -0,0 +1,98 @@
+import React, { useState, useMemo, useEffect } from "react";
+import auth0, { Auth0Result } from "auth0-js";
+import * as Sentry from "@sentry/browser";
+import { AppContext, GeoCoordinates, SessionCacher, AuthService } from "utils";
+import config from "../config";
+
+export interface AuthState {
+  isAuthenticated: boolean;
+  user: {
+    id: string;
+    email: string;
+  };
+  accessTokenObject: {
+    token: string;
+    expiresAt: Date;
+  };
+}
+
+export const defaultAuthObject: AuthState = {
+  isAuthenticated: false,
+  user: {
+    id: "",
+    email: "",
+  },
+  accessTokenObject: {
+    token: "",
+    expiresAt: new Date(1970, 0, 1),
+  },
+};
+
+export const AppProvider = ({
+  children,
+  userLocation,
+}: {
+  children: React.ReactNode;
+  userLocation: GeoCoordinates | null;
+}) => {
+  const authObject = SessionCacher.getAuthObject() ?? defaultAuthObject;
+  const [authState, setAuthState] = useState(authObject);
+  useEffect(() => {
+    // This effect runs after any changes to the AppContext's authState and syncs the changes
+    // to the authObject in sessionStorage.
+    SessionCacher.setAuthObject(authState);
+  }, [authState]);
+
+  const contextValue = useMemo(() => {
+    const authClient = new auth0.WebAuth({
+      audience: config.AUTH0_AUDIENCE,
+      clientID: config.AUTH0_CLIENT_ID,
+      domain: config.AUTH0_DOMAIN,
+      redirectUri: config.AUTH0_REDIRECT_URI,
+      responseType: "token id_token",
+    });
+
+    return {
+      userLocation,
+      authState,
+      setAuthState,
+      authClient,
+    };
+  }, [authState, userLocation]);
```

Comment:
Two things: 1) I think we probably don't want to put the entire thing in the `useMemo()`, and 2) I don't think we actually want to use `useMemo()` at all for this.

The reason we shouldn't put the entire context in the `useMemo()` is because we'll be needlessly invalidating the memoization any time `authState` or `userLocation` changes, but I believe the real thing you're trying to not run multiple times is the `WebAuth()` object. Since it looks like that object's credentials are all static, not influenced by any of the props or state in this component, we should really just be memoizing that and not the other, cheaper to create values.

The reason we shouldn't use `useMemo()` here is because it isn't guaranteed to actually cache at all. If the `WebAuth` object contains any mutable state, we really ought to be storing this in something that is stored globally outside of React, either as a global variable, or possibly with a storage slot created with [`useRef()`](https://react.dev/reference/react/useRef). Initializing and access the global variable/ref would have to be done with `useEffect()`, similar to this example in the docs for `useEffect()`: https://react.dev/reference/react/useEffect#controlling-a-non-react-widget

If it doesn't have any mutable state, then it might be fine to just store it in a `useState()`, where we never call the setter method, so it'll always use the initial value.

Actually, it looks like there's an official auth0-react library, which basically handles all this stuff for us: https://github.com/auth0/auth0-react. It's probably worth using, since it takes care of this state management for us, and it seems to provide nice hooks for us to use, which might even mean that we might not need most of the changes in this file.

Reply:
The one unfortunate issue with `Auth0-React` for our purposes is that it does not support embedded passwordless login; i.e., an embedded form that sends user credentials to Auth0's auth servers without a redirect to an Auth0 interstitial page. Auth0 is really pushing their universal login flow, which involves the user being redirected to an interstitial page under Auth0's domain where they enter email/verification code credentials. That was my reason for not using the React SDK since our designs use an embedded form. 

Below is some discussion on auth0-react and embedded passwordless login. It looks like they have no intention of allowing it. Anyhow, in the meantime, I will look into replacing the useMemo hook with useState or useRef.

- https://github.com/auth0/auth0-spa-js/pull/702#issuecomment-792822354
- https://github.com/auth0/auth0-spa-js/issues/1088#issuecomment-1493887388

Reply:
In hindsight it probably would have been a good bit easier had we chosen the universal flow, but here we are...

Reply:
FWIW, I was racking my brain trying to remember why/how I decided to use `useMemo`. I believe that I lifted it from `Auth0-React`'s implementation of their AppContext object. See [their code here](https://github.com/auth0/auth0-react/blob/c461b11429899a62413371affb1675b13f61cb66/src/auth0-provider.tsx#L270). That said, I'm still looking into doing this without `useMemo`

Reply:
I tried defining the context value with `useState`, but when I do this, the user's authentication state does not propagate downwards after a successful login; this is evident because the "Log Out" button does not appear after auth'ing the user. I have confirmed that the authState is updated in the Context component but not in its children components. I'm not quite sure why this is. My only guess is that there is some kind of weirdness going on given that the `context` value basically a state value that is referencing other state values and that when these other state values update, it does not update for some reason.

For experimental purposes I tried defining the context as a simple const var. Then, I got an interesting linter error: `The 'contextValue' object (at line 74) passed as the value prop to the Context provider (at line 129) changes every render. To fix this consider wrapping it in a useMemo hook.`

This makes me think that maybe `useMemo` is a more common practice for the React AppContext API? This is admittedly getting to the edge of my React knowledge, but I have also seen some discussion online about using `useMemo` in the AppContext provider. In addition to the Auth0 react app using it in their AppContext provider (see my link above).